lgsm/functions/workshop_core.sh

```diff
@@ -0,0 +1,318 @@
+			if [ ! -e "${dst}" ]; then
+				# Finally we can rename the file
+				mv "${src}" "${dst}"
+				# Exit if it fails for any reason
+				local exitcode=$?
+				if [ "${exitcode}" != 0 ]; then
+					fn_print_fail_eol_nl
+					core_exit.sh
+				fi
+			fi
+		done < <(find "${workshopmodsdir}" -depth -name '*[[:upper:]]*' -print0)
+		fn_print_ok_eol_nl
+	fi
+}
+
+# # Copy the mod into serverfiles.
+fn_workshop_copy_destination() {
+	local modid="$1"
+	local modname="$(fn_workshop_get_mod_name ${modid})"
+	if fn_workshop_is_mod_copy_needed ${modid}; then
+		echo "Copying mod ${modname} (${modid})"
+		# If workshop mod exists in installation folder, delete it for clean install
+		if [ -d "${workshopmodsdir}/${modid}" ]; then
+			rm -rf "${workshopmodsdir}/${modid}"
+		fi
+		modsrc="${workshopmodsdldir}/steamapps/workshop/content/${gameappid}/${modid}"
+		cp -fa ${modsrc} ${workshopmodsdir}
+		if [ "${engine}" == "realvirtuality" ]; then
+			modkey="${workshopmodsdldir}/steamapps/workshop/content/${gameappid}/${modid}/keys"
+			if ! [ -d "${modkey}" ]; then
+				modkey="$steamcmd/steamapps/workshop/content/${gameappid}/${modid}/Keys"
+			fi
+			if ! [ -d "${modkey}" ]; then
+				modkey="$steamcmd/steamapps/workshop/content/${gameappid}/${modid}/key"
+			fi
+			if ! [ -d "${modkey}" ]; then
+				modkey="$steamcmd/steamapps/workshop/content/${gameappid}/${modid}/Key"
+			fi
+			if ! [ -d "${modkey}" ]; then
+				echo "Mod ${modname} seems to be missing key folder. Tring to copy key from the main folder."
+				cp -fa "${workshopmodsdir}/${modid}/*.bikey" ${keysdir}
+			else
+				cp -fa ${modkey}/*.bikey ${keysdir}
+			fi
+		fi
+	else
+		echo "Mod ${modname}  is already in mods folder."
+	fi
+}
+
+# ## Directory management.
+
+# Create mods files and directories if it doesn't exist.
+fn_create_workshop_dir() {
+	# Create lgsm data modsdir.
+	if [ ! -d "${workshopmodsdldir}" ]; then
+		echo -en "creating LinuxGSM Steam Workshop data directory ${workshopmodsdldir}..."
+		mkdir -p "${workshopmodsdldir}"
+		exitcode=$?
+		if [ "${exitcode}" != 0 ]; then
+			fn_print_fail_eol_nl
+			fn_script_log_fatal "Creating mod download dir ${workshopmodsdldir}"
+			core_exit.sh
+		else
+			fn_print_ok_eol_nl
+			fn_script_log_pass "Creating mod download dir ${workshopmodsdldir}"
+		fi
+	fi
+	# Create mod install directory.
+	if [ ! -d "${workshopmodsdir}" ]; then
+		echo -en "creating Steam Workshop install directory ${workshopmodsdir}..."
+		mkdir -p "${workshopmodsdir}"
+		exitcode=$?
+		if [ "${exitcode}" != 0 ]; then
+			fn_print_fail_eol_nl
+			fn_script_log_fatal "Creating mod install directory ${workshopmodsdir}"
+			core_exit.sh
+		else
+			fn_print_ok_eol_nl
+			fn_script_log_pass "Creating mod install directory ${workshopmodsdir}"
+		fi
+	fi
+}
+
+# Counts how many mods were installed.
+fn_workshop_count_installed() {
+	if [ -f "${workshopmodsdir}" ]; then
+		installedworkshopmodscount=$(ls -l "${workshopmodsdir}" | grep -c ^d)
```

Comment:
It's almost never a good idea to include ls(1) in a script, because it's typically redundant, inefficient, and unreliable. The shell already supports glob filename pattern matching, and counting results is simple.

For example

```bash
Count=0
for File in /path/to/files/*; {
    (( Count++ ))
}

printf '%d\n' $Count
```

Or, more concisely and more efficiently, albeit with far less flexibility:

```bash
Files=(/path/to/files/*)
printf '%d\n' ${#Files[@]}
```

This last one is an awesome feature that often gets overlooked.

These features are available in at least >= BASH 3.0.

I'm not sure why you're looking for '^d', so I apologise if I've missed a nuanced and obscure detail. Granted, BASH can effortlessly do what grep(1) is doing here as well.

For example:

```bash
Count=0
for File in *; {
    if [[ $File == *^d* ]]; then
        (( Count++ ))
    fi
}

printf '%d\n' $Count
```

Provided you're looking for '^d' being in filenames. By using these features, that's another 2 programs and a subshell you'd take out of the equation.